pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/observables/ObservablesTNCudaOperator.hpp

```diff
@@ -203,49 +257,229 @@ template <class TensorNetT> class ObservableTNCudaOperator {
                         tensor_network.getNumQubits()));
             }
             modes_.emplace_back(modes_per_term);
+        }
 
-            // modes pointer initialization
-            vector1D<const int32_t *> modesPtrPerTerm;
-            for (std::size_t tensor_idx = 0; tensor_idx < modes_.back().size();
-                 tensor_idx++) {
-                modesPtrPerTerm.emplace_back(modes_.back()[tensor_idx].data());
-            }
-            modesPtr_.emplace_back(modesPtrPerTerm);
-
-            // tensor data initialization
-            vector1D<const void *> tensorDataPtrPerTerm_;
-            for (std::size_t tensor_idx = 0; tensor_idx < numTensors;
-                 tensor_idx++) {
-                auto metaData = obs.getMetaData()[term_idx][tensor_idx];
+        if (!var_cal) {
+            for (std::size_t term_idx = 0; term_idx < numObsTerms_;
+                 term_idx++) {
+                auto numTensors = numTensors_[term_idx];
+
+                // Not required for var calculation below
+                //  modes pointer initialization
+                vector1D<const int32_t *> modesPtrPerTerm;
+                for (std::size_t tensor_idx = 0;
+                     tensor_idx < modes_[term_idx].size(); tensor_idx++) {
+                    modesPtrPerTerm.emplace_back(
+                        modes_[term_idx][tensor_idx].data());
+                }
+                modesPtr_.emplace_back(modesPtrPerTerm);
 
-                auto obsName = std::get<0>(metaData);
-                auto param = std::get<1>(metaData);
-                auto hermitianMatrix = std::get<2>(metaData);
-                std::size_t hash_val = 0;
+                // tensor data initialization
+                vector1D<const void *> tensorDataPtrPerTerm_;
+                for (std::size_t tensor_idx = 0; tensor_idx < numTensors;
+                     tensor_idx++) {
+                    auto obsKey =
+                        add_meta_data_(obs.getMetaData()[term_idx][tensor_idx]);
 
-                if (!hermitianMatrix.empty()) {
-                    hash_val = MatrixHasher()(hermitianMatrix);
+                    tensorDataPtrPerTerm_.emplace_back(
+                        get_obs_device_ptr_(obsKey));
                 }
+                tensorDataPtr_.emplace_back(tensorDataPtrPerTerm_);
+            }
+        } else {
+            cublascaller_ = make_shared_cublas_caller();
+
+            numObsTerms2_ = numObsTerms_ * numObsTerms_;
+
+            for (std::size_t term_idx = 0; term_idx < numObsTerms_;
+                 term_idx++) {
+                for (std::size_t term_idy = 0; term_idy < numObsTerms_;
+                     term_idy++) {
+                    auto coeff = cuDoubleComplex{
+                        coeffs_[term_idx].x * coeffs_[term_idy].x, 0.0};
+                    coeffs2_.emplace_back(coeff);
```

Comment:
Is this loop in essence dealing with H ** 2? If so, could we instead have a function which generates H ** 2 and simply call it in the constructor?

Reply:
aha, this is I'm thinking of. Sure, working on it